### ISSUE TITLE
Styles wrapped in certain media query combinations get dropped

### ISSUE THREAD

**Issue by MichaelThessel** [magento/magento2#21005]
### Preconditions (*)

1. 2.3.0

### Steps to reproduce (*)

1. Install 2.3.0
2. Create a child theme that inherits from the Blank theme

### Expected result (*)

I want to change the breakpoint for when the mobile menu is shown from @screen__m to @screen_l.
For that I updated the breakpoints in _navigation.less
```
  //                                                                               
  //  Mobile                                                                       
  //  _____________________________________________                                
                                                                                   
  .media-width(@extremum, @break) when (@extremum = 'max') and (@break = @screen__l) {
```
```
  //                                                                               
  //  Desktop                                                                      
  //  _____________________________________________                                
                                                                                   
  .media-width(@extremum, @break) when (@extremum = 'min') and (@break = @screen__l) {
```

I expect that the CSS wrapped in these media queries shows up in styles-m.css and styles-l.css respectively.

### Actual result (*)

The styles wrapped in:

```
 .media-width(@extremum, @break) when (@extremum = 'max') and (@break = @screen__l) {
```

Simply get dropped and don't show up in any of the style files. I assume that has something to do with how the syles get split between styles-m.css and styles-l.css. Either way no media query combinations should simply disappear.

**Comment by magento-engcom-team** [comment]
Hi @MichaelThessel. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@MichaelThessel do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by DanielRuf** [comment]
You may have to add this rule to the mediaquery settings as this is none of the core rules.

**Comment by DanielRuf** [comment]
To put it simply, blank theme and Magento CSS lib are mobile first.

**Comment by DanielRuf** [comment]
You have to overwrite / extend https://github.com/magento/magento2/blob/2.3-develop/lib/web/css/source/lib/_responsive.less as there is no media-width entry for max-l but min-l.

**Comment by DanielRuf** [comment]
desktop = l.css, mobile = m.css

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @MichaelThessel thank you for you report, I closing this as this is non-issue, feel free to reopen if you had some additional info

**Comment by MichaelThessel** [comment]
Thanks for the response. I had a look at `_responsive.less` before adding this issue. TBH I cannot believe that you think this is acceptable.

* First of all, that would mean one would need to add and maintain an additional entry to `_responsive.less` for every single breakpoint that is not already on the list. It is a quite common practice to add custom breakpoints to your CSS that are basically customized to the individual use case (i.e. breaking text, breaking for specific asset sizes). I have had designs that often required several individual breakpoints per page. I feel like this has been designed by someone who has never done any frontend development. 
* Stuff just being dropped without an warning is not considered an issue? The compilation should fail at least with an error. 
* Could you please point me to the place in the documentation where it specifies that one is limited to the breakpoints that are already specified in `_responsive.less` and that other media queries get simply dropped?

What makes this even more annoying is that I don't even introduce a new breakpoint I simply try to shift the breaking point of the menu up by one level.

`_navigation.less`

From:
```
.media-width(@extremum, @break) when (@extremum = 'max') and (@break = @screen__m) {
.media-width(@extremum, @break) when (@extremum = 'min') and (@break = @screen__m) {
```

To:
```
.media-width(@extremum, @break) when (@extremum = 'max') and (@break = @screen__l) {
.media-width(@extremum, @break) when (@extremum = 'min') and (@break = @screen__l) {
```


I don't seem to be able to open that ticket again. @engcom-backlog-nazar, @DanielRuf

**Comment by DanielRuf** [comment]
> * First of all, that would mean one would need to add and maintain an additional entry to `_responsive.less` for every single breakpoint that is not already on the list.

This is true and as expected.

**Comment by MichaelThessel** [comment]
@DanielRuf I assume stuff being randomly dropped without any warning is "true and as expected" as well?

**Comment by DanielRuf** [comment]
> * Stuff just being dropped without an warning is not considered an issue? The compilation should fail at least with an error.

Please read the frontend guide for adding new breakpoints and media queries.

https://devdocs.magento.com/guides/v2.3/frontend-dev-guide/responsive-web-design/rwd_css.html

https://devdocs.magento.com/guides/v2.3/frontend-dev-guide/responsive-web-design/rwd-breakpoints.html

**Comment by DanielRuf** [comment]
> .media-width(@extremum, @break) when (@extremum = 'max') and (@break = @screen__l)

This is not in the linked file so you have to add it.

**Comment by DanielRuf** [comment]
> @DanielRuf I assume stuff being randomly dropped without any warning is "true and as expected" as well?

This is the less compiler which ignores unknown mixins / mixin variants as these do not output anything.

**Comment by MichaelThessel** [comment]
> This is the less compiler which ignores unknown mixins / mixin variants as these do not output anything.

It doesn't matter why or where it gets dropped without warning. The problem is that it is dropped without warning. Blaming the less compiler is the same as blaming your swim trunks.

**Comment by DanielRuf** [comment]
The docs are at https://github.com/magento/magento2/blob/2.3-develop/lib/web/css/docs/responsive.html

Also see https://codepen.io/DanielRuf/pen/MLOKZa?editors=1000

**Comment by MichaelThessel** [comment]
NM, what a waste of time.

**Comment by MichaelThessel** [comment]
@magento-engcom-team you might want to think about enforcing some community guidelines. This kind of stuff turns people away from Magento and ultimately hurts the platform.

**Comment by DanielRuf** [comment]
Please see https://codepen.io/DanielRuf/pen/MLOKLa?editors=0100 to understand how mixins in Less work.

If there is no signature with the given values, Less will not compile it and also will not throw any error or warninf.

**Comment by DanielRuf** [comment]
> @magento-engcom-team you might want to think about enforcing some community guidelines. This kind of stuff turns people away from Magento and ultimately hurts the platform.

Sorry but what kind of stuff? This is how Magento has worked the whole years and this is by design (mobile first is always with min, not max).

**Comment by DanielRuf** [comment]
Also please carefully read https://devdocs.magento.com/guides/v2.2/frontend-dev-guide/responsive-web-design/rwd_css.html#lib_rwd and https://devdocs.magento.com/guides/v2.2/frontend-dev-guide/responsive-web-design/rwd_css.html#mobile-first

**Comment by DanielRuf** [comment]
https://zellwk.com/blog/how-to-write-mobile-first-css/
https://www.sitepoint.com/introduction-mobile-first-media-queries/

> The mobile-first code above is a great example of responsible implementation through progressive enhancement.

**Comment by DanielRuf** [comment]
And also http://lesscss.org/features/#mixins-feature-mixin-guards-feature

The `when` (`if`) are conditional mixins and are applied only if the conditions evaluate to `true`.

**Comment by MichaelThessel** [comment]
I understand mobile first design and how the splitting of styles is intended to work. What I'm saying is that the implementation is flawed.

* Stuff shouldn't get dropped without a warning. When has this ever been a reasonable practice in software development?
* Maintaining an additional list of where your stuff needs to go for each individual breakpoint you define is unreasonable. Just imagine a site with 100s of custom breakpoints. 

> This is how Magento has worked the whole years and this is by design (mobile first is always with min, not max).

vendor/magento/theme-frontend-blank/web/css/source/_navigation.less (line 51)

**Comment by DanielRuf** [comment]
> * Stuff shouldn't get dropped without a warning.

Please see the codepen. A function / mixin call which does not have the needed `when` values will do nothing which is by design. Less will ignore it and compile the rest.

There is nothing flawed.

**Comment by DanielRuf** [comment]
> * Maintaining an additional list of where your stuff needs to go for each individual breakpoint you define is unreasonable. Just imagine a site with 100s of custom breakpoints.

You only have to extend / overwrite `_responsive.less` once (for every breakpoint and media query / case).

**Comment by DanielRuf** [comment]
> vendor/magento/theme-frontend-blank/web/css/source/_navigation.less (line 51)

We just have thos and another few but still it is mainly mobile first (min). This was probably needed to reset a few rules or because of a specific module.

**Comment by DanielRuf** [comment]
I have added a few comments to make clear how conditional mixins work:
https://codepen.io/DanielRuf/pen/MLOKLa?editors=0100

**Comment by DanielRuf** [comment]
![image](https://user-images.githubusercontent.com/827205/52365623-35cab600-2a48-11e9-980e-5d6fa885f864.png)

![image](https://user-images.githubusercontent.com/827205/52365644-3fecb480-2a48-11e9-8aa2-4c4534463253.png)

**Comment by DanielRuf** [comment]
> * add custom breakpoints to your CSS that are basically customized to the individual use case (i.e. breaking text, breaking for specific asset sizes

Well generally you have one list of breakpoints for all pages and these are device / viewport based (small / phone, medium / tablet, large / desktop, xlarge / desktop retina / big desktop).

**Comment by magudelo62** [comment]
> > .media-width(@extremum, @break) when (@extremum = 'max') and (@break = @screen__l)
> 
> This is not in the linked file so you have to add it.

Hello @DanielRuf, I have a question why does Magento documentation show a media breakpoint that doesn't work? (Tablet breakpoint)

![image](https://user-images.githubusercontent.com/5323922/63506700-52a37400-c4d6-11e9-88b1-bfe252f1b8a3.png)

I found it so confusing!!!

**Comment by DanielRuf** [comment]
Hi @magudelo62,

Please open a new issue in the docs repo.

**Comment by xpoback** [comment]
This is indeed an issue. I've been working with Magento since 8 years and must unfortunately admit that it's hardly getting better.
